### ISSUE TITLE
Error while starting IoT Edge Runtime with x509 DPS provisioning

### ISSUE THREAD

**Issue by payalgaikwad42** [Azure/iotedge#4580]
I have Linux virtual machine 18.04 LTS with Azure IoT Edge 1.1.0 installed on it. I have ports required for the connection opened in the VM as shown in the image. 

![image](https://user-images.githubusercontent.com/42204123/110590075-4f95d480-819d-11eb-8d0b-fb94581cedfb.png)

I have proxy set on my machine in /etc/environment file. 
when I configure Iot edge with x509 certificates with default connect and listen uri I am getting following error: 

> <3>2021-03-10T06:05:21Z [ERR!] - The daemon could not start up successfully: Could not start management service
<3>2021-03-10T06:05:21Z [ERR!] -        caused by: Invalid URL "fd://iotedge.mgmt.socket/": URL could not be parsed as fd number nor fd name
<3>2021-03-10T06:05:21Z [ERR!] -        caused by: Environment variable LISTEN_FDNAMES is set to an invalid value.
<3>2021-03-10T06:05:21Z [ERR!] -        caused by: environment variable not found

> listen:
  management_uri: "fd://iotedge.mgmt.socket"
  workload_uri: "fd://iotedge.socket"

> connect:
  management_uri: "unix:///var/run/iotedge/mgmt.sock"
  workload_uri: "unix:///var/run/iotedge/workload.sock"

when I set listen uri same as connect url this error is gone. But I am facing another error: iotedge.service: Main process exited, code=exited, status=1/FAILURE
I have attached the logs of journalctl -u iotedge -f and iotedge check --verbose

[journalctllogs.txt](https://github.com/Azure/iotedge/files/6113982/journalctllogs.txt)
[iotedgechecklogs.txt](https://github.com/Azure/iotedge/files/6113993/iotedgechecklogs.txt)

**Comment by darobs** [comment]
Hi @payalgaikwad42,

The "listen" settings, i.e.:
```
management_uri: "fd://iotedge.mgmt.socket"
workload_uri: "fd://iotedge.socket"
```
Means these sockets are managed by systemd and are passed by the `LISTEN_FDNAMES` environment variable. 

You can check to see what systemd is doing to create this socket by the status command:
```
systemctl status iotedge.mgmt.socket
```
This is the systemd target that creates the management socket.  This should have been installed with the iotedge package.

If you change the "listen" section to use the "connect" socket, there may be a few problems you could run into:
- `/var/run/iotedge` is a directory that will be deleted on reboot.  It may not exist if the "iotedge.mgmt.socket" target isn't working.
- The file permissions may not be set correctly if iotedged creates it on its own.

Typically, on an Ubuntu 18.04 system, the best solution is to use the "fd:" uris for listening, and ensure the "iotedge.mgmt.socket" and "iotedge.socket" targets are operating correctly.

**Comment by payalgaikwad42** [comment]
> You can check to see what systemd is doing to create this socket by the status command:
> 
> ```
> systemctl status iotedge.mgmt.socket
> ```

@darobs The result of above command is as follow:

> systemctl status iotedge.mgmt.socket
● iotedge.mgmt.socket - Azure IoT Edge daemon management socket
   Loaded: loaded (/lib/systemd/system/iotedge.mgmt.socket; enabled; vendor preset: enabled)
  Drop-In: /etc/systemd/system/iotedge.mgmt.socket.d
           └─override.conf
   Active: active (running) since Mon 2021-03-15 18:30:09 IST; 16h ago
     Docs: man:iotedged(8)
   Listen: /var/run/iotedge/mgmt.sock (Stream)
    Tasks: 0 (limit: 4915)
   CGroup: /system.slice/iotedge.mgmt.socket
Mar 15 18:30:09 ubuntu-virtual-machine systemd[1]: Starting Azure IoT Edge daemon management socket.
Mar 15 18:30:09 ubuntu-virtual-machine systemd[1]: Listening on Azure IoT Edge daemon management socket.


I have kept the default connect and listen uri's. But I am getting resource not available error in iotedge check command to connect to dps and iothub endpoints. 

> × host can connect to and perform TLS handshake with DPS endpoint - Error
    Could not connect to global.azure-devices-provisioning.net : could not resolve hostname
        caused by: failed to lookup address information: Name or service not known
× host can connect to and perform TLS handshake with IoT Hub AMQP port - Error
    Could not connect to gehclcscsidevusea06-iot-hub.azure-devices.net:5671 : could not resolve hostname
        caused by: failed to lookup address information: Name or service not known
× host can connect to and perform TLS handshake with IoT Hub HTTPS / WebSockets port - Error
    Could not connect to gehclcscsidevusea06-iot-hub.azure-devices.net:443 : could not resolve hostname
        caused by: failed to lookup address information: Name or service not known
× host can connect to and perform TLS handshake with IoT Hub MQTT port - Error
    Could not connect to gehclcscsidevusea06-iot-hub.azure-devices.net:8883 : could not resolve hostname
        caused by: failed to lookup address information: Name or service not known
× container on the default network can connect to IoT Hub AMQP port - Error
    Container on the default network could not connect to gehclcscsidevusea06-iot-hub.azure-devices.net:5671
        caused by: docker returned exit code: 1, stderr = One or more errors occurred. (Resource temporarily unavailable)
× container on the default network can connect to IoT Hub HTTPS / WebSockets port - Error
    Container on the default network could not connect to gehclcscsidevusea06-iot-hub.azure-devices.net:443
        caused by: docker returned exit code: 1, stderr = One or more errors occurred. (Resource temporarily unavailable)
× container on the default network can connect to IoT Hub MQTT port - Error
    Container on the default network could not connect to gehclcscsidevusea06-iot-hub.azure-devices.net:8883
        caused by: docker returned exit code: 1, stderr = One or more errors occurred. (Resource temporarily unavailable)
× container on the IoT Edge module network can connect to IoT Hub AMQP port - Error
    Container on the azure-iot-edge network could not connect to gehclcscsidevusea06-iot-hub.azure-devices.net:5671
        caused by: docker returned exit code: 1, stderr = One or more errors occurred. (Resource temporarily unavailable)
× container on the IoT Edge module network can connect to IoT Hub HTTPS / WebSockets port - Error
    Container on the azure-iot-edge network could not connect to gehclcscsidevusea06-iot-hub.azure-devices.net:443
        caused by: docker returned exit code: 1, stderr = One or more errors occurred. (Resource temporarily unavailable)
× container on the IoT Edge module network can connect to IoT Hub MQTT port - Error
    Container on the azure-iot-edge network could not connect to gehclcscsidevusea06-iot-hub.azure-devices.net:8883
        caused by: docker returned exit code: 1, stderr = One or more errors occurred. (Resource temporarily unavailable)

**Comment by payalgaikwad42** [comment]
one thing I noticed is EdgeAgent container runtime is in UTC timezone and my VM machine is in local timezone

**Comment by darobs** [comment]
The edgeAgent container logs in UTC, which is annoying, but useful for coordination with the IoT Hub service, but the container time should be local.

"could not resolve hostname" is a problem on your host network. Are you able to connect to the services with commands like `curl`?

```
curl -v https://global.azure-devices-provisioning.net
```

**Comment by payalgaikwad42** [comment]
> "could not resolve hostname" is a problem on your host network. Are you able to connect to the services with commands like `curl`?
> 
> ```
> curl -v https://global.azure-devices-provisioning.net
> ```
**Output of above command is:**
>  curl -v https://global.azure-devices-provisioning.net
 Rebuilt URL to: https://global.azure-devices-provisioning.net/
   Trying 10.195.54.53...
 TCP_NODELAY set
 Connected to cis-india-pitc-bangalore.corporate.ge.com (10.195.54.53) port 80 (#0)
 allocate connect buffer!
 Establish HTTP proxy tunnel to global.azure-devices-provisioning.net:443
> CONNECT global.azure-devices-provisioning.net:443 HTTP/1.1
> Host: global.azure-devices-provisioning.net:443
> User-Agent: curl/7.58.0
> Proxy-Connection: Keep-Alive
< HTTP/1.1 200 Connection Established
< Proxy-Agent: Zscaler/6.0
 Proxy replied 200 to CONNECT request
 CONNECT phase completed!
 ALPN, offering h2
 ALPN, offering http/1.1
 successfully set certificate verify locations:
   CAfile: /etc/ssl/certs/ca-certificates.crt
  CApath: /etc/ssl/certs
 TLSv1.3 (OUT), TLS handshake, Client hello (1):
 CONNECT phase completed!
 CONNECT phase completed!
 TLSv1.3 (IN), TLS handshake, Server hello (2):
 TLSv1.2 (IN), TLS handshake, Certificate (11):
 TLSv1.2 (IN), TLS handshake, Server key exchange (12):
 TLSv1.2 (IN), TLS handshake, Request CERT (13):
 TLSv1.2 (IN), TLS handshake, Server finished (14):
 TLSv1.2 (OUT), TLS handshake, Certificate (11):
 TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
 TLSv1.2 (OUT), TLS change cipher, Client hello (1):
 TLSv1.2 (OUT), TLS handshake, Finished (20):
 TLSv1.2 (IN), TLS handshake, Finished (20):
 SSL connection using TLSv1.2 / ECDHE-RSA-AES128-GCM-SHA256
 ALPN, server did not agree to a protocol
 Server certificate:
  subject: CN=.azure-devices-provisioning.net
  start date: Nov  9 19:00:43 2020 GMT
  expire date: Nov  9 19:00:43 2021 GMT
  subjectAltName: host "global.azure-devices-provisioning.net" matched cert's ".azure-devices-provisioning.net"
  issuer: C=US; O=Microsoft Corporation; CN=Microsoft RSA TLS CA 01
  SSL certificate verify ok.
> GET / HTTP/1.1
> Host: global.azure-devices-provisioning.net
> User-Agent: curl/7.58.0
> Accept: /
< HTTP/1.1 404 Not Found
< Date: Wed, 17 Mar 2021 06:37:41 GMT
< Content-Length: 0
< x-ms-request-id: 618eb65c-2f4f-4297-8068-48f910e8c061
< Strict-Transport-Security: max-age=31536000; includeSubDomains
 Connection #0 to host cis-india-pitc-bangalore.corporate.ge.com left intact



> "could not resolve hostname"

for this  have localhost 127.0.0.1 and hostname 127.0.0.1 entries in my /etc/hosts file. I have also specified DNS entry ["1.1.1.1"] in /etc/docker/daemon.json file.

**Comment by payalgaikwad42** [comment]
I have added https_proxy in config.yaml file. It seems iotedge runtime is able to connect to DPS endpoint but not to the IoT Hub endpoint:

> curl -v  gehclcscsidevusea06-iot-hub.azure-devices.net:8883
Rebuilt URL to: gehclcscsidevusea06-iot-hub.azure-devices.net:8883/
Trying 10.195.54.52...
TCP_NODELAY set
Connected to PITC-Zscaler-ASPAC-Bangalore3PR.proxy.corporate.ge.com (10.195.54.52) port 80 (#0)
> GET http://gehclcscsidevusea06-iot-hub.azure-devices.net:8883/ HTTP/1.1
> Host: gehclcscsidevusea06-iot-hub.azure-devices.net:8883
> User-Agent: curl/7.58.0
> Accept: */*
> Proxy-Connection: Keep-Alive
>HTTP 1.0, assume close after body
< HTTP/1.0 504 Gateway Timeout
< Server: Zscaler/6.0
< Content-Type: text/html
< Connection: close
><html><head><title>504 Gateway Timeout</title></head>
><body><h1>Gateway Timeout</h1>
><p>Server error - server 20.49.109.129 is unreachable at this moment.<br><br>Please retry the request or contact your >administrator.<br></p>
></body></html>
 >Closing connection 0

**Comment by darobs** [comment]
Hi @payalgaikwad42 

Thank you for your patience. 

1. Please make sure the DNS entry is appropriate for your network.  Especially in corporate networks, some DNS services are blocked. I usually use the DNS service for my internet-facing connection as shown in `resolvectl status`
2. If you need a proxy to connect outside your network, you may need to switch the upstream protocol, We have a [document that describes the work to setup behind a proxy.](https://docs.microsoft.com/azure/iot-edge/how-to-configure-proxy-support?view=iotedge-2018-06). 
3

**Comment by payalgaikwad42** [comment]
@darobs Thank You , I will try this and get back to you.

**Comment by payalgaikwad42** [comment]
1.  There is no resolvectl status command found. 
2.  Getting error for following command: systemd-resolve status

> status: resolve call failed: No appropriate name servers or networks for name found

3.  This is ubuntu VM could you please tell which is the right file to set DNS server?

**Comment by darobs** [comment]
@payalgaikwad42 `resolvectl` is available on Ubuntu 20.04,  It's `systemd-resolve --status` on 18.04 or earlier.

**Comment by darobs** [comment]
Hi @ payalgaikwad42, I think all of your current questions are either answered by the above or the DNS section in [our common error resolutions](https://docs.microsoft.com/azure/iot-edge/troubleshoot-common-errors?view=iotedge-2018-06#edge-agent-module-reports-empty-config-file-and-no-modules-start-on-the-device) page.  Closing this issue.
